File: src/main/java/render/PlaneRenderer.java
package render;

import gl.FragmentShader;
import gl.ShaderProgram;
import gl.VertexShader;
import org.joml.Matrix4f;
import org.joml.Vector3f;
import org.lwjgl.opengl.GL31;
import org.lwjgl.system.MemoryStack;
import util.FloatList;
import world.WorldType;

import java.io.File;
import java.io.IOException;

public class PlaneRenderer {
    private static int _vao;
    private static int _vbo;
    private static FloatList _verts = new FloatList();
    private static int _quads;

    private static final float SIZE = 64.0f;

    private static ShaderProgram _planeShader;
    private static int _planeTrans;
    private static int _planeView;
    private static int _planeProj;
    private static int _planeFog;
    private static int _planeColor;

    private static Vector3f ATMOS_COLOR = new Vector3f();
    private static Vector3f SKY_PLANE = new Vector3f(0.000f, 0.749f, 1.000f);
    private static Vector3f VOID_PLANE = new Vector3f(0.118f, 0.565f, 1.000f);
    private static Vector3f DARK_PLANE = new Vector3f();

    private static Matrix4f _planeTransMat = new Matrix4f();

    private static boolean hasInitialized = false;

    private static MemoryStack _stack = null;

    public static void init() throws IOException {
        if(hasInitialized) return;
        _vao = GL31.glGenVertexArrays();
        GL31.glBindVertexArray(_vao);
        _vbo = GL31.glGenBuffers();
        GL31.glBindBuffer(GL31.GL_ARRAY_BUFFER, _vbo);

        renderPlane(16.0f);
        renderPlane(-16.0f);

        VertexShader planeVert= new VertexShader(new File("shader", "skyplane.vert"));
        FragmentShader planeFrag = new FragmentShader(new File("shader", "skyplane.frag"));
        _planeShader = new ShaderProgram(planeVert, planeFrag);

        _planeShader.use();

        _planeTrans = GL31.glGetUniformLocation(_planeShader.getProgram(), "model");
        _planeView = GL31.glGetUniformLocation(_planeShader.getProgram(), "view");
        _planeProj = GL31.glGetUniformLocation(_planeShader.getProgram(), "proj");
        _planeColor = GL31.glGetUniformLocation(_planeShader.getProgram(), "color");
        _planeFog = GL31.glGetUniformLocation(_planeShader.getProgram(), "fogColor");

        int posAttrib = GL31.glGetAttribLocation(_planeShader.getProgram(), "position");
        GL31.glEnableVertexAttribArray(posAttrib);
        GL31.glVertexAttribPointer(posAttrib, 3, GL31.GL_FLOAT, false, 3*4, 0L);
    }

    public static void setColors(WorldType type) {
        ATMOS_COLOR = type.atmoColor;
        SKY_PLANE = type.skyColor;
        VOID_PLANE = type.voidColor;
    }

    private static void renderPlane(float y) {
        _verts.append(-SIZE);
        _verts.append(y);
        _verts.append(-SIZE);

        _verts.append(y < 0 ? -SIZE : SIZE);
        _verts.append(y);
        _verts.append(y < 0 ? SIZE : -SIZE);

        _verts.append(SIZE);
        _verts.append(y);
        _verts.append(SIZE);

        _verts.append(SIZE);
        _verts.append(y);
        _verts.append(SIZE);

        _verts.append(y < 0 ? SIZE : -SIZE);
        _verts.append(y);
        _verts.append(y < 0 ? -SIZE : SIZE);

        _verts.append(-SIZE);
        _verts.append(y);
        _verts.append(-SIZE);

        _quads = _verts.getLength() / 3;

        GL31.glBufferData(GL31.GL_ARRAY_BUFFER, _verts.getStore(), GL31.GL_DYNAMIC_DRAW);
    }

    public static void draw(Matrix4f view, Matrix4f proj, Vector3f pos, Double pitch, Double yaw) {
        _planeShader.use();
        GL31.glDisable(GL31.GL_TEXTURE);
        GL31.glDepthMask(false);
        _stack = null;
        try {
            _stack = MemoryStack.stackPush();
            GL31.glBindVertexArray(_vao);
            GL31.glBindBuffer(GL31.GL_ARRAY_BUFFER, _vbo);
            _planeTransMat = new Matrix4f()
                    .translate(pos);
            GL31.glUniformMatrix4fv(_planeTrans, false, _planeTransMat.get(_stack.mallocFloat(16)));
            GL31.glUniformMatrix4fv(_planeView, false, view.get(_stack.mallocFloat(16)));
            GL31.glUniformMatrix4fv(_planeProj, false, proj.get(_stack.mallocFloat(16)));
            GL31.glUniform3fv(_planeColor, SKY_PLANE.get(_stack.mallocFloat(3)));
            GL31.glUniform3fv(_planeFog, ATMOS_COLOR.get(_stack.mallocFloat(3)));
            GL31.glDrawArrays(GL31.GL_TRIANGLES, 0, _quads/2);
            if(pos.y < 62) {
                GL31.glUniform3fv(_planeColor, DARK_PLANE.get(_stack.mallocFloat(3)));
            } else {
                GL31.glUniform3fv(_planeColor, VOID_PLANE.get(_stack.mallocFloat(3)));
            }
            GL31.glDrawArrays(GL31.GL_TRIANGLES, _quads/2, _quads);

        } finally {
            _stack.pop();
        }
        GL31.glDepthMask(true);
        GL31.glEnable(GL31.GL_TEXTURE);
    }
}

File: src/main/kotlin/render/PlaneRenderer.kt
package render

import gl.FragmentShader
import gl.ShaderProgram
import gl.VertexShader
import org.joml.Matrix4f
import org.joml.Vector3f
import org.lwjgl.opengl.GL31.*
import org.lwjgl.system.MemoryStack
import world.WorldType
import java.io.File

object PlaneRenderer {
    var vao: Int = 0
    var vbo: Int = 0
    val verts = mutableListOf<Float>()
    var quads: Int = 0

    val size = 64.0f

    val planeShader: ShaderProgram
    var planeTrans: Int
    var planeView: Int
    var planeProj: Int
    var planeFog: Int
    var planeColor: Int

    var ATMOS_COLOR: Vector3f = Vector3f()
    var SKY_PLANE: Vector3f = Vector3f(0.000f, 0.749f, 1.000f)
    var VOID_PLANE: Vector3f = Vector3f(0.118f, 0.565f, 1.000f)
    val DARK_PLANE: Vector3f = Vector3f()

    var planeTransMat = Matrix4f()

    init {
        vao = glGenVertexArrays()
        glBindVertexArray(vao)
        vbo = glGenBuffers()
        glBindBuffer(GL_ARRAY_BUFFER, vbo)

        renderPlane(16.0f)
        renderPlane(-16.0f)

        var planeVert= VertexShader(File("shader", "skyplane.vert"))
        var planeFrag = FragmentShader(File("shader", "skyplane.frag"))
        planeShader = ShaderProgram(planeVert, planeFrag)

        planeShader.use()

        planeTrans = glGetUniformLocation(planeShader.getProgram(), "model")
        planeView = glGetUniformLocation(planeShader.getProgram(), "view")
        planeProj = glGetUniformLocation(planeShader.getProgram(), "proj")
        planeColor = glGetUniformLocation(planeShader.getProgram(), "color")
        planeFog = glGetUniformLocation(planeShader.getProgram(), "fogColor")

        var posAttrib = glGetAttribLocation(planeShader.getProgram(), "position")
        glEnableVertexAttribArray(posAttrib)
        glVertexAttribPointer(posAttrib, 3, GL_FLOAT, false, 3*4, 0L)
    }

    fun setColors(type: WorldType) {
        ATMOS_COLOR = type.atmoColor
        SKY_PLANE = type.skyColor
        VOID_PLANE = type.voidColor
    }

    fun renderPlane(y: Float) {
        verts.add(-size)
        verts.add(y)
        verts.add(-size)

        verts.add(if(y < 0) -size else size)
        verts.add(y)
        verts.add(if(y < 0) size else -size)

        verts.add(size)
        verts.add(y)
        verts.add(size)

        verts.add(size)
        verts.add(y)
        verts.add(size)

        verts.add(if(y < 0) size else -size)
        verts.add(y)
        verts.add(if(y < 0) -size else size)

        verts.add(-size)
        verts.add(y)
        verts.add(-size)

        quads = verts.size / 3

        glBufferData(GL_ARRAY_BUFFER, verts.toFloatArray(), GL_DYNAMIC_DRAW)
    }

    fun draw(view: Matrix4f, proj: Matrix4f, pos: Vector3f, pitch: Double, yaw: Double) {
        planeShader.use()
        glDisable(GL_TEXTURE)
        glDepthMask(false)
        var stack: MemoryStack? = null
        try {
            stack = MemoryStack.stackPush()
            glBindVertexArray(vao)
            glBindBuffer(GL_ARRAY_BUFFER, vbo)
            planeTransMat = Matrix4f()
                .translate(pos)
            glUniformMatrix4fv(planeTrans, false, planeTransMat.get(stack.mallocFloat(16)))
            glUniformMatrix4fv(planeView, false, view.get(stack.mallocFloat(16)))
            glUniformMatrix4fv(planeProj, false, proj.get(stack.mallocFloat(16)))
            glUniform3fv(planeColor, SKY_PLANE.get(stack.mallocFloat(3)))
            glUniform3fv(planeFog, ATMOS_COLOR.get(stack.mallocFloat(3)))
            glDrawArrays(GL_TRIANGLES, 0, quads/2)
            if(pos.y < 62) {
                glUniform3fv(planeColor, DARK_PLANE.get(stack.mallocFloat(3)))
            } else {
                glUniform3fv(planeColor, VOID_PLANE.get(stack.mallocFloat(3)))
            }
            glDrawArrays(GL_TRIANGLES, quads/2, quads)

        } finally {
            stack?.pop()
        }
        glDepthMask(true)
        glEnable(GL_TEXTURE)
    }
}